hstream/src/HStream/Server/Handler.hs

```diff
@@ -432,63 +425,80 @@ checkSubscriptions timeout ServerContext{..} =  do
-        _             -> return $ Left "Unknown error"
-
-    doSubscribe :: S.LDClient -> SubscribedReaders -> TVar (Map TL.Text Int64) -> S.StreamId -> IO (ServerResponse 'Normal Subscription)
-    doSubscribe client sReaders subscribeHp streamName = do
-      reader <- S.newLDRsmCkpReader scLDClient (textToCBytes $ TL.toStrict subscriptionSubscriptionId)
-        S.checkpointStoreLogID 5000 1 Nothing 10
+subscribeHandler
+  :: ServerContext
+  -> ServerRequest 'Normal SubscribeRequest SubscribeResponse
+  -> IO (ServerResponse 'Normal SubscribeResponse)
+subscribeHandler ServerContext{..} (ServerNormalRequest _metadata SubscribeRequest{..}) = defaultExceptionHandle $ do
+  (reader, Subscription{..}) <- doCheck subscribedReaders subscribeRequestSubscriptionId
+  let sName = transToStreamName $ TL.toStrict subscriptionStreamName
+  doSubscribe scLDClient reader sName
+  where
+    -- Confirm subscriptionId exist, and the reader status should be Released, then update the reader status to Occupied.
+    doCheck :: SubscribedReaders -> TL.Text -> IO (S.LDSyncCkpReader, Subscription)
+    doCheck readers subscriptionId = atomically $ do
+      hm <- readTVar readers
+      case HM.lookup subscriptionId hm of
+        Just (ReaderMap rd sId Released) -> do
+          let newMap = HM.insert subscriptionId (ReaderMap rd sId Occupied) hm
+          writeTVar readers newMap
+          return (rd, sId)
+        Just (ReaderMap _ _ Occupied) -> throwSTM SubscriptionIdOccupied
+        Just None -> error "reader status shouldn't be None in subscribeHandler, but get None."
+        Nothing -> throwSTM SubscriptionIdNotFound
+
+    doSubscribe :: S.LDClient -> S.LDSyncCkpReader -> S.StreamId -> IO (ServerResponse 'Normal SubscribeResponse)
+    doSubscribe client reader streamName = do
       logId <- S.getUnderlyingLogId client streamName
-      startLSN <- getStartLSN (fromJust subscriptionOffset) client logId
-      res <- try $ S.ckpReaderStartReading reader logId startLSN S.LSN_MAX
-      case res of
-        Right _ -> do
-          currentTime <- getCurrentTimestamp
-          let readerMap = ReaderMap reader subscription Occupied
-          atomically $ do
-            updateSubscribedReaders sReaders subscriptionSubscriptionId readerMap
-            modifyTVar' subscribeHp $ Map.insert subscriptionSubscriptionId currentTime
-        Left _ -> atomically $ deleteSubscribedReaders sReaders subscriptionSubscriptionId
-      eitherToResponse res subscription
+      S.startReadingFromCheckpoint reader logId S.LSN_MAX
+      currentTime <- getCurrentTimestamp
+      atomically $ modifyTVar' subscribeHeap $ Map.insert subscribeRequestSubscriptionId currentTime
+      returnResp $ SubscribeResponse subscribeRequestSubscriptionId
+      `catch` \e -> do
+        atomically $ updateReaderStatus subscribedReaders Released subscribeRequestSubscriptionId
+        returnErrResp StatusInternal (StatusDetails $ BS.pack . displayException $ (e :: SomeException))
```

Comment:
use a modified version of `defaultExceptionHandle`

Reply:
fixed